FastSimulation/TrackingRecHitProducer/src/PixelTemplateSmearerBase.cc

```diff
@@ -0,0 +1,1001 @@
+    bool hasBigPixelInX = rectPixelTopology->containsBigPixelInX( firstPixelInX, lastPixelInX );
+    bool hasBigPixelInY = rectPixelTopology->containsBigPixelInY( firstPixelInY, lastPixelInY );
+
+    //Variables for SiPixelTemplate pixel hit error output
+    float sigmay, sigmax, sy1, sy2, sx1, sx2;  
+    templ.temperrors(
+        templateId, cotalpha, cotbeta, nqbin,          // inputs
+        sigmay, sigmax, sy1, sy2, sx1, sx2        // outputs
+    );
+
+    if(edge)
+    {
+        if(edgex && !edgey)
+        {
+            theErrorX = 23.0*microntocm;
+            theErrorY = 39.0*microntocm;
+        }
+        else if(!edgex && edgey)
+        {
+            theErrorX = 24.0*microntocm;
+            theErrorY = 96.0*microntocm;
+        }
+        else
+        {
+            theErrorX = 31.0*microntocm;
+            theErrorY = 90.0*microntocm;
+        }
+    }  
+    else 
+    {
+        if (singlex)
+        {
+            if (hitbigx)
+            {
+                theErrorX = sx2*microntocm;
+            }
+            else
+            {
+                theErrorX = sx1*microntocm;
+            }
+        }
+        else
+        {
+            theErrorX = sigmax*microntocm;
+        }
+        if (singley)
+        {
+            if (hitbigy)
+            {
+                theErrorY = sy2*microntocm;
+            }
+            else
+            {
+                theErrorY = sy1*microntocm;
+            }
+        }
+        else 
+        {
+            theErrorY = sigmay*microntocm;
+        }
+    }
+
+
+    //add misalignment error
+    const TrackerGeomDet* misalignmentDetUnit = getMisalignedGeometry().idToDet(detUnit->geographicalId());
+    const LocalError& alignmentError = misalignmentDetUnit->localAlignmentError();
+    if (alignmentError.valid())
+    {
+        theError = LocalError( 
+            theErrorX*theErrorX+alignmentError.xx(), 
+            alignmentError.xy(), 
+            theErrorY*theErrorY+alignmentError.yy()
+        );
+    }
+    else
+    {
+        theError = LocalError( 
+            theErrorX*theErrorX, 
+            0.0,
+            theErrorY*theErrorY
+        );
+    }
+
+    // Local Error is 2D: (xx,xy,yy), square of sigma in first an third position 
+    // as for resolution matrix
+    // Generate position
+    // get resolution histograms
+    int cotalphaHistBin = (int)( ( cotalpha - rescotAlpha_binMin ) / rescotAlpha_binWidth + 1 );
+    int cotbetaHistBin  = (int)( ( cotbeta  - rescotBeta_binMin )  / rescotBeta_binWidth + 1 );
+    // protection against out-of-range (undeflows and overflows)
+    if (cotalphaHistBin < 1) cotalphaHistBin = 1; 
+    if (cotbetaHistBin  < 1) cotbetaHistBin  = 1; 
+    if (cotalphaHistBin > (int)rescotAlpha_binN) cotalphaHistBin = (int)rescotAlpha_binN; 
+    if (cotbetaHistBin  > (int)rescotBeta_binN) cotbetaHistBin  = (int)rescotBeta_binN; 
+    //
+    unsigned int theXHistN;
+    unsigned int theYHistN;
+
+    if (!isForward)
```

Comment:
The generation of the key to find backt the right histogram in the template map is implemented twice.
Once here and once in the initialisation of
PixelForwardTemplateSmearerPlugin and PixelBarrelTemplateSmearerPlugin.
One should write a function, e.g. something like
getSmearingTemplateKey(
    cotalphaHistBin,
    cotbetaHistBin,
    isForward,
    edge,
    singlex,
    hitbigx,
    hasBigPixelInX,
    singley,
    hitbigy,
    hasBigPixelInY)